build_runner/lib/src/generate/heartbeat.dart

```diff
@@ -10,75 +10,127 @@ import '../logging/human_readable_duration.dart';
       throw new StateError('HeartbeatLogger already started');
     }
     _totalWatch = new Stopwatch()..start();
     _intervalWatch = new Stopwatch()..start();
-    _resetHeartbeat();
-    _listener = Logger.root.onRecord.listen((_) => _resetHeartbeat());
-    _timer = new Timer.periodic(
-        const Duration(milliseconds: 100), _logIfDurationIsMet);
+    ping();
+    _timer = new Timer.periodic(checkInterval, _checkDuration);
   }
 
   /// Stops this heartbeat logger, must already be started.
+  ///
+  /// This method can be overridden to add additional logic for cleanup
+  /// purposes, but you must call `super.stop()`.
   void stop() {
-    if (_totalWatch == null ||
-        _intervalWatch == null ||
-        _listener == null ||
-        _timer == null) {
+    if (_totalWatch == null || _intervalWatch == null || _timer == null) {
       throw new StateError('HeartbeatLogger was never started');
     }
     _totalWatch.stop();
     _totalWatch = null;
     _intervalWatch.stop();
     _intervalWatch = null;
-    _listener.cancel();
-    _listener = null;
     _timer.cancel();
     _timer = null;
   }
 
-  /// Resets the current [_intervalWatch].
-  void _resetHeartbeat() {
-    _intervalWatch.reset();
+  void _checkDuration(_) {
+    if (_intervalWatch.elapsed < waitDuration) return;
+    onTimeout(_intervalWatch.elapsed);
+    ping();
   }
+}
 
-  /// Logs a heartbeat message if [_intervalWatch] has been running for
-  /// [waitDuration] or more.
-  void _logIfDurationIsMet(_) {
-    if (_intervalWatch.elapsed < waitDuration) return;
+/// Watches [Logger.root] and if there are no logs for [waitDuration] then it
+/// will log a heartbeat message with the current elapsed time since [start] was
+/// originally invoked.
+class HeartbeatLogger extends Heartbeat {
+  StreamSubscription<LogRecord> _listener;
+
+  /// Will be invoked with each original log message and the returned value will
+  /// be logged instead.
+  final String Function(String original) transformLog;
 
-    var formattedTime = humanReadable(_totalWatch.elapsed);
+  HeartbeatLogger(
+      {Duration checkInterval, Duration waitDuration, this.transformLog})
+      : super(checkInterval: checkInterval, waitDuration: waitDuration);
+
+  /// Start listening to logs.
+  @override
+  void start() {
+    super.start();
+    _listener = Logger.root.onRecord.listen((_) => ping());
+  }
+
+  /// Stops listenting to the logger;
+  @override
+  void stop() {
+    _listener.cancel();
+    _listener = null;
+  }
+
+  /// Logs a heartbeat message if we reach the timeout.
+  @override
+  void onTimeout(Duration elapsed) {
+    var formattedTime = humanReadable(elapsed);
     var message = '$formattedTime elapsed';
     if (transformLog != null) {
       message = transformLog(message);
     }
     _logger.info(message);
-    _resetHeartbeat();
+  }
+}
+
+class HungActionsHeartbeat extends Heartbeat {
+  /// An arbitraty function that will return a description of pending actions.
```

Comment:
How about `/// Returns a description of pending actions`

Reply:
done